### ISSUE TITLE
Linker script implementation

### ISSUE THREAD

**Issue by biemster** [biemster/st17h66_FindMy#1]
The Lenze SDK is written for the ARMCC compiler, and uses Keil uVision. Linking is done using a scatter file `scatter_load.sct`. GCC is using a linker script, which is not compatible with the scatter file. The process of converting the uVision uvproject was done using https://github.com/phodina/ProjectConverter which provides an example GCC linker script, but this has to be modified by hand.

Some useful resources on how to do this are:
https://interrupt.memfault.com/blog/how-to-write-linker-scripts-for-firmware
https://www.renesas.com/kr/en/document/apn/b-024-porting-keil-uvision-project-gnu-chain-tool
https://github.com/adamheinrich/cm-makefile

Also uVision can be configured to use GCC (this maybe generates a linker script):
https://developer.arm.com/documentation/101407/0537/Creating-Applications/Tips-and-Tricks/GNU-C-Compiler-Support
https://developer.arm.com/downloads/-/gnu-rm

**Comment by biemster** [comment]
Some object files in `lib/rf.lib` seem to have an invalid `sh_info` value, and the GNU linker is stumbling over them:
```
:~/temp/Lenze/ST17H65_SDK_3.1.1.2_lowPower_beacon/lib/rflib$ readelf -s patch.o > /dev/null
readelf: Warning: local symbol 748 found at index >= .symtab's sh_info value of 418
```
because `readelf` is writing this to `stderr`.

**Comment by biemster** [comment]
Patching `patch.o` and `rf_phy_driver.o` with a hex editor actually fixes this issue. Next problem:
```
simpleBlePeripheral.elf: hidden symbol 'osal_memcmp' isn't defined
```

**Comment by biemster** [comment]
Seems like an issue in GCC not supporting the **exact** CPU architecture:
https://stackoverflow.com/a/16688192/11545161
So I see two possible solutions atm:
1. Convince the linker that it is fine to link
2. Get a toolchain with support for `thumb/cortex-m0;@mthumb@mcpu=cortex-m0`

**Comment by biemster** [comment]
https://github.com/biemster/Lenze_libs should be exactly the source for the libs, as `ST17H65_SDK_3.1.1.2_lowPower_beacon/lib/rflib/phy_sec_ext.o` contains the string `PHY+62XXPLUS0504` which is the same `s_company_id` found in `phy_sec_ext.c` in that repository! It might be from an older version of the SDK though, so I might need to dig up some old SDKs. I already have 3.0.9, and I remember seeing 3.0.6 somewhere as well..

**Comment by biemster** [comment]
More people are trying something similar: https://github.com/AloyseTech/PHY6202_GCC

**Comment by biemster** [comment]
So I can get it to compile fully to a `hex` file using GCC in uVision on Windows, good first step! I'm not fully sure though if everything got linked correctly, but I do have all the ingredients now.
One minus is that the `hex` is too big to fit on the chip, LeKit complains that the XIP part is too large. I tried to discard unused objects using `-ffunction-sections`, `-fdata-sections`, and then linker flag `--gc-sections`, but then uVision reports that gcc does not know about `-gc-sections`. So Keil seems to inject some linker options into the compiler command.. :(

**Comment by biemster** [comment]
It's compiling now to a 90kB hex under linux using GCC, but when flashing this to the chip nothing seems to happen..

**Comment by biemster** [comment]
The ch6121 looks very similar to the 17h66 and has an GCC SDK: https://github.com/diskman88/PHY6212

**Comment by biemster** [comment]
I've started https://github.com/biemster/st17h66_blinky to first try to convert an absolute minimum project to GCC

**Comment by biemster** [comment]
Blinky compiles and runs now with GCC!

**Comment by biemster** [comment]
https://github.com/biemster/st17h66_RF now also compiles on GCC, including sleep mode! As this is much simpler code, this repo will be rewritten to use the RF only approach build there. Since that compiles on GCC now, this can be closed.

**Comment by wangwillian0** [comment]
> https://github.com/biemster/st17h66_RF now also compiles on GCC, including sleep mode! As this is much simpler code, this repo will be rewritten to use the RF only approach build there. Since that compiles on GCC now, this can be closed.

Hi, I just tried the current GCC build and it doesn't work (no BLE activity), do you think it would be easy to refactor this repo using that approach? I'm quite busy these days but maybe I can help with one thing or another

**Comment by biemster** [comment]
It should work, I definitely got BLE activity there also on the binaries built by GCC. Getting this repo with the whole osal and bluetooth stack should be quite straightforward to port to GCC too, although the linker script and specifically the XIP stuff will need some work.

**Comment by steve-m** [comment]
I recently played around with this and added the FindMy specific code, and I had the tag appear on the FindMy map:
https://github.com/steve-m/st17h66_RF

**Comment by wangwillian0** [comment]
@steve-m I just tried it and the beacon doesn't seem to be broadcasting any packet, did you confirm it by bluetooth? maybe the problem is with my local setup

**Comment by steve-m** [comment]
Yes, that seems likely. I am starting an LE scan with the bluetoothctl command "scan le", and filter the packets in Wireshark to match the MAC address of my beacon. I can see the periodic advertisements there.

**Comment by wangwillian0** [comment]
I tested again under ubuntu packages using docker (I was originally with Manjaro), compiled with make and got the same result. I also tried https://github.com/biemster/st17h66_blinky and nothing blinked at all, which makes me think I'm with a slightly different version of hardware. Building with the ARM compiler from Keil works fine.

**Comment by wangwillian0** [comment]
Ok, my mistake, I was using the wrong flash_st17h66.py 🤦. I can confirm your repository is working, it also periodically makes a sound and prints debug messages to serial. Thank you!

**Comment by biemster** [comment]
Great work! Keep in mind though that enabling the UART will decrease your battery life from close to a year to a week or two.

**Comment by steve-m** [comment]
Yes, that should be disabled again. I needed it for debugging, because when there is no BLE output and no UART, you don't know what goes wrong ;)

**Comment by Philip2809** [comment]
I don't really know where to write this as https://github.com/steve-m/st17h66_RF does not have an issues page but because it was discussed here i will write it here. 

I have installed this firmware and it does indeed work, however I am wondering why it does not show up on nRF connect mobile app while the firmware from this repo does? I used bluetoothctl with `scan le` and there it showed up, any ideas?

I used gcc with steve-m/rf and was unable to get keil to work with this repo

**Comment by biemster** [comment]
All this stuff ended up in an almost finished state, what did you do exactly? the _RF code does indeed compile with GCC, but I did not implement the FindMy advertisement there if I recall correctly. the _FindMy repo has the correct advertisement, but only compiles with Keil in the current state.

it is actually trivially easy to add the FindMy advertisement to the _RF repo, but somehow life got in between...

**Comment by Philip2809** [comment]
I realize now I wrote that in a confusing way. I got the _RF to compile and send findmy advertisements with gcc used the steve-m fork. However i was unable to compile _FindMy, but as you say its not very hard to add the FindMy stuff to the _RF. What i am wondering is why the precompiled _FindMy you provide in the FindMy repo shows up in nrf connect as a FindMy device but when using the _RF that i compiled it does not show on nrf but i can scan for it with my computer with bluetoothctl. I have only used the default key so far, will generate a new one and try it, bring it with me and see if it makes any reports, tho uart is enabled so it will kill my battery. Want to add a flag to turn it on or off in compile if possible

Edit: what version of keil etc?
